tests/test_io.py

```diff
@@ -0,0 +1,201 @@
+#!/usr/bin/env python3
+from Bio import SeqIO
+from Bio.Seq import Seq
+from Bio.SeqRecord import SeqRecord
+import bz2
+import gzip
+import lzma
+from pathlib import Path
+import pytest
+import random
+import sys
+
+# make sure we get the local version of modules
+sys.path.insert(0, (str(Path(__file__).parent.parent.parent)))
+from augur.io import read_sequences, write_sequences
+
+
+def random_seq(k):
+    """Generate a single random sequence of nucleotides of length k.
+    """
+    return "".join(random.choices(("A","T","G","C"), k=k))
+
+def generate_sequences(n, k=10):
+    """Generate n random sequences of length k.
+    """
+    return (
+        SeqRecord(Seq(random_seq(k)), id=f"SEQ_{i}")
+        for i in range(1, n + 1)
+    )
+
+@pytest.fixture
+def sequences():
+    return list(generate_sequences(3))
+
+@pytest.fixture
+def sequences_generator():
+    return generate_sequences(3)
+
+@pytest.fixture
+def fasta_filename(tmpdir, sequences):
+    filename = str(tmpdir / "sequences.fasta")
+    SeqIO.write(sequences, filename, "fasta")
+    return filename
+
+@pytest.fixture
+def additional_fasta_filename(tmpdir, sequences):
+    filename = str(tmpdir / "additional_sequences.fasta")
+    SeqIO.write(sequences, filename, "fasta")
+    return filename
+
+@pytest.fixture
+def gzip_fasta_filename(tmpdir, sequences):
+    filename = str(tmpdir / "sequences.fasta.gz")
+
+    with gzip.open(filename, "wt") as oh:
+        SeqIO.write(sequences, oh, "fasta")
+
+    return filename
+
+@pytest.fixture
+def bzip2_fasta_filename(tmpdir, sequences):
+    filename = str(tmpdir / "sequences.fasta.bz2")
+
+    with bz2.open(filename, "wt") as oh:
+        SeqIO.write(sequences, oh, "fasta")
+
+    return filename
+
+@pytest.fixture
+def lzma_fasta_filename(tmpdir, sequences):
+    filename = str(tmpdir / "sequences.fasta.xz")
+
+    with lzma.open(filename, "wt") as oh:
+        SeqIO.write(sequences, oh, "fasta")
+
+    return filename
+
+@pytest.fixture
+def genbank_reference():
+    return "tests/builds/zika/config/zika_outgroup.gb"
+
+
+class TestReadSequences:
+    def test_read_sequences_from_single_file(self, fasta_filename):
+        sequences = read_sequences(fasta_filename, "fasta")
+        assert len([sequence for sequence in sequences]) == 3
```

Comment:
FWIW substituting `len([sequence for sequence in sequences])` with `len(list(sequences))` is a bit simpler to read and possibly [more efficient than doing a list comprehension](https://stackoverflow.com/a/50645935/6287798) on the `sequences` generator.

Reply:
There are 5 other instances of this but I'm not going to comment on all of them.